src/main/java/org/folio/service/orders/flows/update/open/OpenCompositeOrderFlowValidator.java

```diff
@@ -122,41 +128,79 @@ public Future<Void> checkFundLocationRestrictions(List<CompositePoLine> poLines,
 
-    // Checking fund location against validLocations in POL to identify restricted locations
-    // if there is one, will be stored to put in error as parameter
-    Set<String> restrictedLocations = funds.stream()
-      .filter(Fund::getRestrictByLocations)
-      .flatMap(fund -> fund.getLocationIds().stream())
-      .filter(locationId -> !validLocations.contains(locationId))
-      .collect(Collectors.toSet());
-
-    if (restrictedLocations.isEmpty()) {
+    if (fundsWithRestrictedLocations.isEmpty()) {
+      logger.info("validateLocationRestrictions:: Funds is not restricted by locations");
       return Future.succeededFuture();
     }
 
-    String poLineId = poLine.getId();
-    logger.error("For POL {} locations {} are restricted to be used by all funds", poLineId, restrictedLocations);
-    List<Parameter> parameters = List.of(
-      new Parameter().withKey("poLineId").withValue(poLineId),
-      new Parameter().withKey("poLineNumber").withValue(poLine.getPoLineNumber()),
-      new Parameter().withKey("restrictedLocations").withValue(restrictedLocations.toString())
-    );
-    return Future.failedFuture(new HttpException(422, ErrorCodes.FUND_LOCATION_RESTRICTION_VIOLATION, parameters));
+    return extractRestrictedLocations(poLine, fundsWithRestrictedLocations, requestContext)
+      .compose(restrictedLocations -> {
+        if (restrictedLocations.isEmpty()) {
+          logger.info("validateLocationRestrictions:: restricted locations is not found for poLineId '{}'", poLine.getId());
+          return Future.succeededFuture();
+        }
+
+        String poLineId = poLine.getId();
+        logger.error("For POL {} locations {} are restricted to be used by all funds", poLineId, restrictedLocations);
+        List<Parameter> parameters = List.of(
+          new Parameter().withKey("poLineId").withValue(poLineId),
+          new Parameter().withKey("poLineNumber").withValue(poLine.getPoLineNumber()),
+          new Parameter().withKey("restrictedLocations").withValue(restrictedLocations.toString())
+        );
+        return Future.failedFuture(new HttpException(422, ErrorCodes.FUND_LOCATION_RESTRICTION_VIOLATION, parameters));
+      });
+  }
+
+  private Future<Set<String>> extractRestrictedLocations(CompositePoLine poLine, Set<Fund> fundsWithRestrictedLocations, RequestContext requestContext) {
+    Set<String> validLocationIds = poLine.getLocations().stream().map(Location::getLocationId).filter(Objects::nonNull).collect(Collectors.toSet());
+    List<String> holdingIds = poLine.getLocations().stream().map(Location::getHoldingId).filter(Objects::nonNull).toList();
+
+    return getPermanentLocationIdsFromHoldings(holdingIds, requestContext)
+      .map(permanentLocationIds -> {
+        logger.info("extractRestrictedLocations:: fundsWithRestrictedLocations: {} is being checked against permanentLocationIds: {} and validLocations: {}", fundsWithRestrictedLocations, permanentLocationIds, validLocationIds);
```

Comment:
fundsWithRestrictedLocations it is Set of Fund objects, and Fund object does not have to string overriden
Also can move params to the next line to avoid this long line

Reply:
Thanks, improved this part